### PR TITLE
Feature/gcn normalization layer

### PR DESCRIPTION
**Pull request by sktzwhj**
Hi @adocherty and @youph 

This PR implements the GCN preprocessing as a Keras layer (only for tf backend though). Please have a look and comments and suggestions are appreciated. Thanks.

Huijun

### CODE REVIEW

File: stellargraph/utils/saliency_maps/preprocessing_layer.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            )
+        )
+        # Add self loops.
+        adj = adj + tf.eye(tf.shape(adj)[0])
+        # Normalization
+        rowsum = tf.reduce_sum(adj, 1)
+        p = tf.fill(tf.shape(rowsum), -0.5)
+        d_inv_sqrt = tf.pow(rowsum, p)
+        self.d_inv_sqrt_var = tf.assign(self.d_inv_sqrt_var, d_inv_sqrt)
+        tf.assign(
```
[the remark this answers is not in the document]
Comment (adocherty) [reply stellargraph/utils/saliency_maps/preprocessing_layer.py]: I agree, and if this is the case we shouldn't need the Tensorflow variable `self.d_inv_sqrt_var` anymore?

File: stellargraph/layer/gcn.py [hunk 2]
```diff
@@ -304,6 +305,10 @@ def __init__(
 
         # Check if the generator is producing a sparse matrix
         self.use_sparse = generator.use_sparse
+        if self.method == "none":
```
[the remark this answers is not in the document]
Comment (youph) [reply stellargraph/layer/gcn.py]: `method="tf"` is yet another option to explain to users. Right now they should be very careful to choose the correct method when creating the generator. Should we go back to having a really simple generator that does not apply any transforms to the adjacency matrix, and all the transforms are done in the layers? Or we can not add self-loops in GAT layers anymore?

File: stellargraph/utils/saliency_maps/preprocessing_layer.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            adj
+            + tf.multiply(
+                adj_T, tf.where(adj_T > adj, tf.ones_like(adj), tf.zeros_like(adj))
+            )
+            - tf.multiply(
+                adj, tf.where(adj_T > adj, tf.ones_like(adj), tf.zeros_like(adj))
+            )
+        )
+        # Add self loops.
+        adj = adj + tf.eye(tf.shape(adj)[0])
```
[the remark this answers is not in the document]
Comment (youph) [reply stellargraph/utils/saliency_maps/preprocessing_layer.py]: If we could, this would be a way to add self-loops in GAT layer, and then we can do all transforms in the models/layers, not in generators!

File: stellargraph/layer/gcn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (adocherty) [reply stellargraph/layer/gcn.py]: Let's get this issue through and then we can discuss this. I tend to agree that we should simplify the potential transforms and let the model handle this.

File: stellargraph/utils/saliency_maps/preprocessing_layer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adocherty) [reply stellargraph/utils/saliency_maps/preprocessing_layer.py]: This isn't going to work for sparse matrices unfortunately. However, adding a diagonal matrix with the diagonal elements subtracted, as above, could work!

File: stellargraph/utils/saliency_maps/preprocessing_layer.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    """
+    This class implements the pre-processing of adjacency matrices in GCN. We implement it in tensorflow so that
+    while computing the saliency maps, we are able to calculate the gradients in an end-to-end way. 
+    We currently only support this for tensorflow backend.
+
+    Args:
+    output_dim (int pair): The output shape of the pre-processed adjacency matrix.
+    """
+
+    def __init__(self, output_dim=(2708, 2708), **kwargs):
```
[the remark this answers is not in the document]
Comment (sktzwhj) [reply stellargraph/utils/saliency_maps/preprocessing_layer.py]: Yes, you are correct. Fixed now.

File: stellargraph/utils/saliency_maps/preprocessing_layer.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        adj_T = tf.transpose(adj)
+        adj = (
+            adj
+            + tf.multiply(
+                adj_T, tf.where(adj_T > adj, tf.ones_like(adj), tf.zeros_like(adj))
+            )
+            - tf.multiply(
+                adj, tf.where(adj_T > adj, tf.ones_like(adj), tf.zeros_like(adj))
+            )
+        )
```
[the remark this answers is not in the document]
Comment (sktzwhj) [reply stellargraph/utils/saliency_maps/preprocessing_layer.py]: @adocherty I did that as this line 'A = A + A.T.multiply(A.T > A) - A.multiply(A.T > A)
' exists in the previous code. I'd like to keep it to make it consistent.

File: stellargraph/utils/saliency_maps/preprocessing_layer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sktzwhj) [reply stellargraph/utils/saliency_maps/preprocessing_layer.py]: Yes. 

adj = adj + tf.linalg.diag(tf.ones(adj.shape[0]) - tf.diag_part(adj)) does the job.

File: stellargraph/layer/gcn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sktzwhj) [reply stellargraph/layer/gcn.py]: Keep it unchanged. Leave it for future decisions.

File: stellargraph/utils/saliency_maps/preprocessing_layer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sktzwhj) [reply stellargraph/utils/saliency_maps/preprocessing_layer.py]: Yes, we do not need it anymore.

**Comment by adocherty** [comment]
> I think this looks good, but I also think we should follow this up immediately (as a new issue) by removing the adjacency matrix transforms in the FullBatchNodeGenerator class. @adocherty what do you think?

If we do that then we loose support for SGC which is defined as a transform of the matrix in the generator.  We will also loose the ability to do the high-order Chebyshev approximations in the GCN paper.  We could re-implement SGC by using a similar custom input layer to the one for GCN in this issue.  However, we can't implement the Chebyshev approximations (due to requirements for SVD on the matrix).

So we have a decision to make:
1. Deprecate adjacency matrix pre-processing in generator and re-implement SGC as pre-processing layer.
2. Default to pre-processing layer for GCN but also support current generator pre-processing for SGC.
3. Keep GCN pre-processing layer as optional and default to current generator pre-processing. 

I tend to favour 2.

**Comment by sktzwhj** [comment]
Merge this branch for now and we will come back soon and make the decision Andrew suggested soon.
